contracts/utils/MemoryUint.sol

```diff
@@ -0,0 +1,681 @@
+        SharedMemory memory mem_,
+        Uint512 memory a_,
+        Uint512 memory b_,
+        Uint512 memory m_
+    ) internal view returns (Uint512 memory r_) {
+        _checkMemory(mem_, 64);
+
+        MemoryStack.StackValue memory aExt_ = _extend(mem_, a_.data);
+        MemoryStack.StackValue memory bExt_ = _extend(mem_, b_.data);
+
+        MemoryStack.StackValue memory sum_ = _add(mem_, aExt_, bExt_);
+
+        r_ = Uint512(_mod(mem_, sum_, m_.data));
+
+        _destruct(mem_, aExt_, _StackType._EXT_UINT);
+        _destruct(mem_, bExt_, _StackType._EXT_UINT);
+        _destruct(mem_, sum_, _StackType._EXT_UINT);
+    }
+
+    function modsub(
+        SharedMemory memory mem_,
+        Uint512 memory a_,
+        Uint512 memory b_,
+        Uint512 memory m_
+    ) internal view returns (Uint512 memory r_) {
+        _checkMemory(mem_, 64);
+
+        int cmp_ = _cmp(mem_, a_.data, b_.data);
+
+        MemoryStack.StackValue memory diff_ = cmp_ >= 0
+            ? _sub(mem_, a_.data, b_.data)
+            : _sub(mem_, b_.data, a_.data);
+        MemoryStack.StackValue memory modDiff_ = _mod(mem_, diff_, m_.data);
+
+        _destruct(mem_, diff_, _StackType._UINT);
+
+        if (cmp_ >= 0) {
+            return Uint512(modDiff_);
+        }
+
+        r_ = Uint512(_sub(mem_, m_.data, modDiff_));
+
+        _destruct(mem_, modDiff_, _StackType._UINT);
+    }
+
+    function mod(
+        SharedMemory memory mem_,
+        Uint512 memory a_,
+        Uint512 memory m_
+    ) private view returns (Uint512 memory r_) {
+        _checkMemory(mem_, 64);
+
+        return Uint512(_mod(mem_, a_.data, m_.data));
+    }
+
+    function mul(
+        SharedMemory memory mem_,
+        Uint512 memory a_,
+        Uint512 memory b_
+    ) internal view returns (Uint512 memory r_) {
+        _checkMemory(mem_, 64);
+
+        MemoryStack.StackValue memory rExt_ = _mul(mem_, a_.data, b_.data);
+
+        r_ = Uint512(_cut(mem_, rExt_));
+
+        _destruct(mem_, rExt_, _StackType._EXT_UINT);
+    }
+
+    function modmul(
+        SharedMemory memory mem_,
+        Uint512 memory a_,
+        Uint512 memory b_,
+        Uint512 memory m_
+    ) internal view returns (Uint512 memory r_) {
+        _checkMemory(mem_, 64);
+
+        MemoryStack.StackValue memory rExt_ = _mul(mem_, a_.data, b_.data);
+
+        r_ = Uint512(_mod(mem_, rExt_, m_.data));
+
+        _destruct(mem_, rExt_, _StackType._EXT_UINT);
+    }
+
+    function _modinv(
+        SharedMemory memory mem_,
+        MemoryStack.StackValue memory a_,
+        MemoryStack.StackValue memory m_
+    ) internal view returns (MemoryStack.StackValue memory r_) {
+        MemoryStack.StackValue memory two_ = _newUint(mem_, 2);
+
+        require(_cmp(mem_, m_, two_) >= 0, "MU: invalid modulus");
+
+        MemoryStack.StackValue memory exponent_ = _sub(mem_, m_, two_);
+
+        _destruct(mem_, two_, _StackType._UINT);
+
+        r_ = _modexp(mem_, a_, exponent_, m_);
```

Comment:
This only works if `m_` is prime. Shall we check the correctness of the computation?